source/mir/ndslice/sparse.d

```diff
@@ -0,0 +1,623 @@
+import std.meta;
+
+import mir.ndslice.internal;
+import mir.ndslice.slice;
+
+/++
+Sparse tensors represented in Dictionary of Keys (DOK) format.
+
+Params:
+    N = dimension count
+    lengths = list of dimension lengths
+Returns:
+    `N`-dimensional slice composed of indexes
+See_also: $(LREF Sparse)
++/
+Sparse!(Lengths.length, T) sparse(T, Lengths...)(Lengths lengths)
+    if (allSatisfy!(isIndex, Lengths))
+{
+    return .sparse!(T, Lengths.length)([lengths]);
+}
+
+/// ditto
+Sparse!(N, T) sparse(T, size_t N)(auto ref size_t[N] lengths)
+{
+    import mir.ndslice.slice: sliced;
+    T[size_t] table;
+    table[0] = 0;
+    table.remove(0);
+    assert(table !is null);
+    with (typeof(return)) return SparseMap!T(table).sliced(lengths);
+}
+
+///
+pure unittest
+{
+    auto slice = sparse!double(2, 3);
+    slice[0][] = 1;
+    slice[0, 1] = 2;
+    --slice[0, 0];
+    slice[1, 2] += 4;
+
+    assert(slice == [[0, 2, 1], [0, 0, 4]]);
+
+    import std.range.primitives: isRandomAccessRange;
+    static assert(isRandomAccessRange!(Sparse!(2, double)));
+
+	import mir.ndslice.slice: Slice, DeepElementType;
+    static assert(is(Sparse!(2, double) : Slice!(2, Range), Range));
+    static assert(is(DeepElementType!(Sparse!(2, double)) == double));
+}
+
+/++
+Sparse Slice in Dictionary of Keys (DOK) format.
++/
+alias Sparse(size_t N, T) = Slice!(N, SparseMap!T);
+
+/++
+SparseMap is a range, which is used internally by $(LREF Sparse).
++/
+struct SparseMap(T)
+{
+	///
+	T[size_t] table;
+
+	///
+	auto save() @property
+	{
+		return this;
+	}
+
+	///
+	T opIndex(size_t index)
+	{
+		static if (isScalarType!T)
+			return table.get(index, cast(T)0);
+		else
+			return table.get(index, null);
+	}
+
+	///
+	T opIndexAssign(T value, size_t index)
+	{
+		static if (isScalarType!T)
+		{
+			if (value != 0)
+			{
+				table[index] = value;
+			}
+		}
+		else
+		{
+			if (value !is null)
+			{
+				table[index] = value;
+			}
+		}
+		return value;
+	}
+
+	///
+	T opIndexUnary(string op)(size_t index)
+		if (op == `++` || op == `--`)
+	{
+		mixin (`auto value = ` ~ op ~ `table[index];`);
+		static if (isScalarType!T)
+		{
+			if (value == 0)
+			{
+				table.remove(index);
+			}
+		}
+		else
+		{
+			if (value is null)
+			{
+				table.remove(index);
+			}
+		}
+		return value;
+ 	}
+
+ 	///
+	T opIndexOpAssign(string op)(T value, size_t index)
+		if (op == `+` || op == `-`)
+	{
+		mixin (`value = table[index] ` ~ op ~ `= value;`); // this works
+		static if (isScalarType!T)
+		{
+			if (value == 0)
+			{
+				table.remove(index);
+			}
+		}
+		else
+		{
+			if (value is null)
+			{
+				table.remove(index);
+			}
+		}
+		return value;
+    }
+}
+
+/++
++/
+struct CoordinateValue(size_t N, T)
+{
+	///
+	size_t[N] index;
+	
+	///
+	T value;
+
+	///
+	sizediff_t opCmp()(auto ref const typeof(this) rht) const
+	{
+		return cmpCoo(this.index, rht.index);
+	}
+}
+
+private sizediff_t cmpCoo(size_t N)(const auto ref size_t[N] a, const auto ref size_t[N] b)
```

Comment:
you only use this once (three lines above) - thus why not directly putting it there?